Mathlib/RingTheory/DividedPowers/Padic.lean

```diff
@@ -0,0 +1,236 @@
+Authors: Antoine Chambert-Loir, María Inés de Frutos-Fernández
+-/
+
+import Mathlib.NumberTheory.Padics.PadicIntegers
+import Mathlib.RingTheory.DividedPowers.RatAlgebra
+
+/-! # Divided powers on ℤ_[p]
+
+Given a divided power algebra `(B, J, δ)` and an injective ring morphism `f : A →+* B`, if `I` is
+an `A`-ideal such that `I.map f = J` and such that for all `n : ℕ`, `x ∈ I`, the preimage of
+`hJ.dpow n (f x)` under `f` belongs to `I`, we get an induced divided power structure on `I`.
+
+We specialize this construction to the coercion map `ℤ_[p] →+* ℚ_[p]` to get a divided power
+structure on the ideal `(p) ⊆ ℤ_[p]`. This divided power structure is given by the family of maps
+`fun n x ↦ x^n / n!`.
+
+TODO: If `K` is a `p`-adic local field with ring of integers `R` and uniformizer `π` such that
+`p = u * π^e` for some unit `u`, then the ideal `(π) ⊆ R` has divided powers if and only if
+`e ≤ p - 1`.
+
+-/
+
+namespace PadicInt
+
+open DividedPowers DividedPowers.OfInvertibleFactorial Nat Ring
+
+variable (p : ℕ) [hp : Fact p.Prime]
+
+section Factorial
+
+theorem sub_one_mul_padicValNat_factorial_lt_of_ne_zero {n : ℕ} (hn : n ≠ 0) :
+    (p - 1) * padicValNat p n.factorial < n := by
+  have hpn : ((p - 1 : ℕ) : ℝ) * (padicValNat p n ! : ℝ) = n - (p.digits n).sum := by
+    rw [← cast_mul, sub_one_mul_padicValNat_factorial n, cast_sub (digit_sum_le p n)]
+  rw [← Nat.cast_lt (α := ℝ), Nat.cast_mul, hpn]
+  have hnil : p.digits n ≠ [] := Nat.digits_ne_nil_iff_ne_zero.mpr hn
+  apply sub_lt_self _
+  exact_mod_cast Nat.sum_pos_iff_exists_pos.mpr ⟨(p.digits n).getLast hnil,
+    List.getLast_mem hnil, Nat.pos_of_ne_zero (Nat.getLast_digit_ne_zero p hn)⟩
+
+theorem padicValNat_factorial_lt_of_ne_zero {n : ℕ} (hn : n ≠ 0) :
+    padicValNat p n.factorial < n := by
+  apply lt_of_le_of_lt _ (sub_one_mul_padicValNat_factorial_lt_of_ne_zero p hn)
+  conv_lhs => rw [← one_mul (padicValNat p n !)]
+  gcongr
+  exact le_sub_one_of_lt (Nat.Prime.one_lt hp.elim)
+
+theorem padicValNat_factorial_le (n : ℕ) : padicValNat p n.factorial ≤ n := by
+  by_cases hn : n = 0
+  · simp [hn]
+  · exact le_of_lt (padicValNat_factorial_lt_of_ne_zero p hn)
+
+end Factorial
+
+section Injective
+
+open Function
+
+variable {A B : Type*} [CommSemiring A] [CommSemiring B] (I : Ideal A) (J : Ideal B)
+
+/-- Given a divided power algebra `(B, J, δ)` and an injective ring morphism `f : A →+* B`, if `I`
+is an `A`-ideal such that `I.map f = J` and such that for all `n : ℕ`, `x ∈ I`, the preimage of
+`hJ.dpow n (f x)` under `f` belongs to `I`, this is the induced divided power structure on `I`. -/
+noncomputable def dividedPowers_of_injective (f : A →+* B) (hf : Injective f)
+    (hJ : DividedPowers J) (hIJ : I.map f = J)
+    (hmem : ∀ (n : ℕ) {x : A} (_ : x ∈ I), ∃ (y : A) (_ : n ≠ 0 → y ∈ I), f y = hJ.dpow n (f x)) :
+    DividedPowers I where
+  dpow n x := open Classical in if hx : x ∈ I then f.toFun.invFun (hJ.dpow n (f x)) else 0
```

Comment:
If you add `[DecidablePred (· ∈ I)]` to the arguments, you can drop the `open Classical in`, and you get for free an implementation which is computable when the underlying ring is sufficiently nice; this might be useful for something, although I'm not completely sure what :-)

Reply:
(PS. Actually this is probably wrong, the main source of non-computability comes from `invFun`, not from the `if ... then` structure. Probably just ignore this.)